### PR TITLE
updates the Kptfile merger so that it handles delete setters correctly.

### PR DESCRIPTION
**Pull request by jijiew**
This PR updates the Kptfile merger so that when the setters deleted, updated Kptfile won't add it back again.

See related PRs on delete setters:
https://github.com/kubernetes-sigs/kustomize/pull/2553

### CODE REVIEW

File: internal/kptfile/pkgfile.go [hunk 1]
```diff
@@ -175,6 +191,33 @@ func shouldSkipCopy(updatedDef, localDef, originalDef *yaml.MapNode, key string)
...
+func shouldRemoveValue (updatedDef, localDef, originalDef *yaml.MapNode, key string) bool  {
+	localVal := localDef.Value.Field(key)
+	originalVal := originalDef.Value.Field(key)
+	updatedVal := updatedDef.Value.Field(key)
+
+	if originalVal == nil || updatedVal == nil {
+		return false
+	}
+
+	originalValStr, err := originalDef.Value.Field(key).Value.String()
```
Comment (phanimarupaka) [inline internal/kptfile/pkgfile.go:204]: you can replace originalDef.Value.Field(key) with originalVal?

Comment (jijiew) [reply internal/kptfile/pkgfile.go]: Done.

**Comment by phanimarupaka** [comment]
@jijiew I think we can merge 2 sections([section1](https://github.com/GoogleContainerTools/kpt/blob/0feffc565395334103c5745b55dbd4a937179abc/internal/kptfile/pkgfile.go#L124), [section2](https://github.com/GoogleContainerTools/kpt/blob/0feffc565395334103c5745b55dbd4a937179abc/internal/kptfile/pkgfile.go#L139)) into 1. You may club/refactor shouldSkipCopy and shouldRemoveValue for this purpose. Essentially, the new method should take localDef, updatedDef and originalDef and return the resultDef which is written back.

**Comment by jijiew** [comment]
> Jijie Wei I think we can merge 2 sections([section1](https://github.com/GoogleContainerTools/kpt/blob/0feffc565395334103c5745b55dbd4a937179abc/internal/kptfile/pkgfile.go#L124), [section2](https://github.com/GoogleContainerTools/kpt/blob/0feffc565395334103c5745b55dbd4a937179abc/internal/kptfile/pkgfile.go#L139)) into 1. You may club/refactor shouldSkipCopy and shouldRemoveValue for this purpose. Essentially, the new method should take localDef, updatedDef and originalDef and return the resultDef which is written back.

Ack. & Done.

File: internal/kptfile/pkgfile.go [hunk 2]
```diff
@@ -121,18 +121,41 @@ func (updatedKf *KptFile) MergeOpenAPI(localKf, originalKf KptFile) error {
 	}
 	oriDef := original.Field("definitions")
 
-	err = localDef.Value.VisitFields(func(node *yaml.MapNode) error {
-		key := node.Key.YNode().Value
-		if shouldSkipCopy(updatedDef, localDef, oriDef, key) {
-			return nil
+	localKeys, _ := localDef.Value.Fields()
```
Comment (phanimarupaka) [inline internal/kptfile/pkgfile.go:124]: The file is not lint checked. Run gofmt -s -w internal/kptfile/pkgfile.go to format the file. Wondering how it passed the checks.

File: internal/kptfile/pkgfile.go [hunk 3]
```diff
@@ -121,18 +121,41 @@ func (updatedKf *KptFile) MergeOpenAPI(localKf, originalKf KptFile) error {
...
+	updatedKeys, _ := updatedDef.Value.Fields()
+    keys := append(updatedKeys, localKeys...)
+
+	unique := make(map[string]bool, len(keys))
```
Comment (phanimarupaka) [inline internal/kptfile/pkgfile.go:128]: Please abstract this algorithm into new method and explain the flow commenting the steps and method definition. It looks good from the first look but more explanation is helpful.

File: internal/kptfile/pkgfile.go [hunk 4]
```diff
@@ -121,18 +121,41 @@ func (updatedKf *KptFile) MergeOpenAPI(localKf, originalKf KptFile) error {
...
+	for i:= 0; i<len(keys); i++ {
```
Comment (phanimarupaka) [inline internal/kptfile/pkgfile.go:129]: for _, key : range keys is more go like.

File: internal/kptfile/pkgfile.go [hunk 5]
```diff
@@ -121,18 +121,41 @@ func (updatedKf *KptFile) MergeOpenAPI(localKf, originalKf KptFile) error {
 	}
 	oriDef := original.Field("definitions")
 
-	err = localDef.Value.VisitFields(func(node *yaml.MapNode) error {
-		key := node.Key.YNode().Value
-		if shouldSkipCopy(updatedDef, localDef, oriDef, key) {
-			return nil
+	localKeys, _ := localDef.Value.Fields()
+	updatedKeys, _ := updatedDef.Value.Fields()
```
Comment (phanimarupaka) [inline internal/kptfile/pkgfile.go:125]: Check for error and return error if any.

File: internal/kptfile/pkgfile.go (hunk 2 above)
Comment (jijiew) [reply internal/kptfile/pkgfile.go]: Done.

File: internal/kptfile/pkgfile.go (hunk 3 above)
Comment (jijiew) [reply internal/kptfile/pkgfile.go:152]: Done.

File: internal/kptfile/pkgfile.go (hunk 4 above)
Comment (jijiew) [reply internal/kptfile/pkgfile.go]: Done.

**Pull request merged**